scene/resources/curve.cpp

```diff
@@ -1630,114 +1698,150 @@ Vector3 Curve3D::sample_baked(real_t p_offset, bool p_cubic) const {
+	Curve3D::Interval interval = _find_interval(p_offset);
+	return _sample_baked(interval, p_cubic);
+}
+
+Transform3D Curve3D::sample_baked_with_rotation(real_t p_offset, bool p_cubic, bool p_apply_tilt) const {
+	if (baked_cache_dirty) {
+		_bake();
 	}
 
-	if (idx == count - 1) {
-		return p_apply_tilt ? r[idx].rotated((rp[idx] - rp[idx - 1]).normalized(), rt[idx]) : r[idx];
+	// Validate: Curve may not have baked points.
+	const int point_count = baked_point_cache.size();
+	ERR_FAIL_COND_V_MSG(point_count == 0, Transform3D(), "No points in Curve3D.");
+
+	if (point_count == 1) {
+		Transform3D t;
+		t.origin = baked_point_cache.get(0);
+		ERR_FAIL_V_MSG(t, "Only 1 point in Curve3D.");
 	}
 
-	real_t offset_begin = baked_dist_cache[idx];
-	real_t offset_end = baked_dist_cache[idx + 1];
+	p_offset = CLAMP(p_offset, 0.0, get_baked_length()); // PathFollower implement wrapping logic.
 
-	real_t idx_interval = offset_end - offset_begin;
-	ERR_FAIL_COND_V_MSG(p_offset < offset_begin || p_offset > offset_end, Vector3(0, 1, 0), "Couldn't find baked segment.");
+	// 0. Find interval for all sampling steps.
+	Curve3D::Interval interval = _find_interval(p_offset);
 
-	real_t frac = (p_offset - offset_begin) / idx_interval;
+	// 1. Sample position.
+	Vector3 pos = _sample_baked(interval, p_cubic);
+
+	// 2. Sample rotation frame.
+	Basis frame = _sample_posture(interval, p_apply_tilt);
+
+	return Transform3D(frame, pos);
+}
+
+real_t Curve3D::sample_baked_tilt(real_t p_offset) const {
+	if (baked_cache_dirty) {
+		_bake();
+	}
 
-	Vector3 forward = (rp[idx + 1] - rp[idx]).normalized();
-	Vector3 up = r[idx];
-	Vector3 up1 = r[idx + 1];
+	// Validate: Curve may not have baked tilts.
+	int pc = baked_tilt_cache.size();
+	ERR_FAIL_COND_V_MSG(pc == 0, 0, "No tilts in Curve3D.");
 
-	if (p_apply_tilt) {
-		up.rotate(forward, rt[idx]);
-		up1.rotate(idx + 2 >= count ? forward : (rp[idx + 2] - rp[idx + 1]).normalized(), rt[idx + 1]);
+	if (pc == 1) {
+		return baked_tilt_cache.get(0);
 	}
 
-	Vector3 axis = up.cross(up1);
+	p_offset = CLAMP(p_offset, 0.0, get_baked_length()); // PathFollower implement wrapping logic
 
-	if (axis.length_squared() < CMP_EPSILON2) {
-		axis = forward;
-	} else {
-		axis.normalize();
+	Curve3D::Interval interval = _find_interval(p_offset);
+	return _sample_baked_tilt(interval);
+}
+
+Vector3 Curve3D::sample_baked_up_vector(real_t p_offset, bool p_apply_tilt) const {
+	if (baked_cache_dirty) {
+		_bake();
+	}
+
+	// Validate: Curve may not have baked up vectors.
+	ERR_FAIL_COND_V_MSG(!up_vector_enabled, Vector3(0, 1, 0), "No up vectors in Curve3D.");
+
+	int count = baked_up_vector_cache.size();
+	if (count == 1) {
+		return baked_up_vector_cache.get(0);
 	}
 
-	return up.rotated(axis, up.angle_to(up1) * frac);
+	p_offset = CLAMP(p_offset, 0.0, get_baked_length()); // PathFollower implement wrapping logic.
+
+	Curve3D::Interval interval = _find_interval(p_offset);
+	return _sample_posture(interval, p_apply_tilt).get_column(1);
```

Comment:
I don't know whether `sample_baked_up_vector()` always needs `forward_vector`. `_sample_posture()` does spherical interpolation of the basis created from the baked `up_vector` and unbaked `forward_vector` inside it, but it seems to me that this is beyond the role of the function `sample_baked_up_vector()`.

Rather, since the only use of `sample_baked_up_vector()` is for paths in `CSGShape`, perhaps we can remove this and keep only `sample_baked_with_rotation()`. But that include unnecessary processing, so I still think it is better to rename `sample_baked_up_vector()` to a function called `sample_up_vector()`.

How do you think? @reduz

Reply:
> created from the baked up_vector and unbaked forward_vector inside it

to not bake the forward_vector is just an optimization, because it is cheap to calculate on the fly. If performance is indeed an issue, baking cache for forward_vector can be add back easily.  

> since the only use of sample_baked_up_vector() is for paths in CSGShape

Another possibility is to remove `sample_baked_up_vector()` altogether, and let `CSGShape` use the rotation frame from `sample_baked_with_rotation()` directly. 

Currently the `CSGShape` path calculate its own rotation frame
https://github.com/godotengine/godot/blob/324106b3dfd7a70c283f5adf17d53cfc48c8cdae/modules/csg/csg_shape.cpp#L1963
with the same method as `sample_baked_with_rotation()` (`looking_at(tangent, up)`). So using rotation frame directly from `sample_baked_with_rotation()` is equivalent right now, but  have advantages for future: Let CSG path object benefit from improvements on Curve3D for free, such as better tilt, bug fixes, performance optimization

Reply:
So I suggested in my comment above that CSGShape use `sample_baked_with_rotation()` and remove `sample_baked_up_vector()`. But it include a process `sample_baked()` which is not needed, so I more recommended renaming it to `sample_up_vector()'.

In any case, the name `sample_baked_up_vector()` seems strange for me because it is sampling something unbaked by `sample_posture()` in it.

The final decision on whether to rename it or not is left to @reduz or another experienced contributor.